Fix for #62 and nested regions example

hroemer opened the pull request:
I created a PR for #62 and additionally provided an example for nested regions.

jrbenny35 commented:
Looks like the travis build failed. Can you take a look at it?

davehunt commented on docs/examples/nested_regions.py:
```diff
@@ -0,0 +1,32 @@
+
+
+class Menu(Region):
+    _root_locator = (By.CLASS_NAME, 'menu')
```
You're passing the root element when instantiating these regions, so `_root_locator` is unecessary, and unused.

davehunt commented on docs/examples/nested_regions.py:
```diff
@@ -0,0 +1,32 @@
+
+
+class Entry(Region):
+    _root_locator = (By.CLASS_NAME, 'entry')
```
Again, the root element is being passed on instantiation, so `_root_locator` is not needed. The `_root_locator` is useful when you have a unique region such as a header or footer and know the precise locator ahead of time. If a `_root_locator` is provided, and no root element is specified on instantiation, then the root element is located whenever it's needed using the locator. This can also help to avoid stale element exceptions with more dynamic pages, as the element is re-located whenever it's needed.

hroemer replied:
Thanks for the explanation, I dropped the mentioned `_root_locator` occurrences and updated the according properties.

hroemer commented on docs/examples/repeated_regions.html:
```diff
@@ -0,0 +1,14 @@
+    <div class="name">Item2</div>
+    <div class="name">Item3</div>
+    <div class="name">Item4</div>
+</div>
```
I still do not understand the intention here!?
My example unit test passes with the HTML snippet and the provided page object:
```python
def test_pypom_repeated_regions(selenium: WebDriver):
    url = 'file:///{}'.format(Path(__file__).parent.joinpath("./repeated_regions.html").resolve())
    page = Results(selenium, base_url=url)
    page.open()
    assert len(page.results) == 4
    assert page.results[1].name == "Item2"
```
Thus it is not required to wrap each `item` in a `result` container.
My understanding was that there is a `results` element (might have been a region as well) on the page that contains a list of repeated `Result` items modeled as `Regions` here.

Could you please elaborate on the example you provided in the user guide before?

hroemer commented:
I still do not understand the intention here!?
My example unit test passes with the HTML snippet and the provided page object:
```python
def test_pypom_repeated_regions(selenium: WebDriver):
    url = 'file:///{}'.format(Path(__file__).parent.joinpath("./repeated_regions.html").resolve())
    page = Results(selenium, base_url=url)
    page.open()
    assert len(page.results) == 4
    assert page.results[1].name == "Item2"
```
Thus it is not required to wrap each `item` in a `result` container.
My understanding was that there is a `results` element (might have been a region as well) on the page that contains a list of repeated `Result` items modeled as `Regions` here.

Could you please elaborate on the example you provided in the user guide before?

davehunt commented:
> Could you please elaborate on the example you provided in the user guide before?

Sorry, coming back to this after 10 days I've lost the context. I'll comment on the current state of the pull request and hopefully make it clearer.

davehunt commented on docs/examples/nested_regions.py:
```diff
@@ -0,0 +1,31 @@
+
+    @property
+    def entries(self):
+        items = self.find_elements(*Entry._root_locator)
```
You shouldn't use `_root_locator` with repeated regions. I would have an `_entries_locator` as a class variable. By defining a `_root_locator`, the behaviour will be different and you may get unexpected elements.

hroemer replied:
OK, I understand why this would lead to problems. Does that imply that a `_root_locator` should be able to uniquely identify an element inside a page or region?

hroemer replied:
OK, just read your last comment about uniqueness.

davehunt commented on docs/examples/nested_regions.py:
```diff
@@ -0,0 +1,31 @@
+    @property
+    def entries(self):
+        items = self.find_elements(*Entry._root_locator)
+        return [Entry(self.page, item) for item in items]
```
I prefer to put this all on one line:

```python
return [Entry(self.page, item) for el in self.find_elements(*self._entries_locator)]
```

When I do define a variable for the elements I tend to name it `els`:

```python
els = self.find_elements(*self._entries_locator)
return [Entry(self.page, item) for el in els]
```

hroemer replied:
The one-liner absolutely makes sense. I did split the list comprehension into two because I am still struggling with the `_root_locator` which is a "private" attribute here that shouldn't be accessed from outer scope anyway.

davehunt commented on docs/examples/repeated_regions.html:
```diff
@@ -0,0 +1,14 @@
+    <div class="item">Item2</div>
+    <div class="item">Item3</div>
+    <div class="item">Item4</div>
+</div>
```
I think the issue is that this HTML is either misleading or insufficient for the example. There appears (to me) to be a single "result", and within that, multiple "items". A related page object might have `page.results` which returns a list of `Result` regions (a list of one), and each of those could have an `items` property.

It's possible that we could clear up the confusion by renaming the classes. If each "item" is in fact a result, then I would renamed the "result" class to something that suggests it's a container, such as "results" or even "content", and then rename "item" to "result".

davehunt commented on docs/examples/repeated_regions.py:
```diff
@@ -0,0 +1,19 @@
+
+    @property
+    def results(self):
+        results_element = self.find_element(*self._result_locator)
```
There's really no need for the `results_element` in this example. If we just want all results then our `_results_locator` should locate them all. With the existing HTML this could be:

```python
_result_locator = (By.CSS_SELECTOR, '.result.item')
```

davehunt commented on docs/examples/repeated_regions.py:
```diff
@@ -0,0 +1,19 @@
+        return [self.Result(self, el) for el in items]
+
+    class Result(Region):
+        _root_locator = (By.CLASS_NAME, 'item')
```
Again, `_root_locator` is more for regions that are unique on the page. By defining it for repeated regions I would expect issues to occur.

hroemer commented:
Thanks for your clarification and recommendations. I will clean up the PR in the next couple of days.

hroemer commented:
After re-reading your initial section about regions again with your comments from above in mind, I think I finally understood your approach. It was all there before, I just couldn't grasp your intention in the example :-).

I reworked the PR and added another detail to the `Result` region, here's the augmented sample unit test:

```python
def test_pypom_repeated_regions(selenium: WebDriver):
    url = 'file:///{}'.format(Path(__file__).parent.joinpath("./repeated_regions.html").resolve())
    page = Results(selenium, base_url=url)
    page.open()
    assert 4 == len(page.results)
    assert "Result 2" == page.results[1].name
    assert "/detail/2" in page.results[1].detail_link
```
I hope it meets your expectations now. Thanks for your patience!

davehunt commented:
Thanks @hroemer this is a great improvement to the docs! 👍
